Sprint 3/step 1

AJIEKCEY opened the pull request.

AJIEKCEY committed:
sprint-3/step-1
Сверстать страницы регистрации, логина, востановления пароля, профиля.
Добавить маршрутизацию.

AJIEKCEY committed:
sprint-3/step-1
Добавить слайсы для авторизации, регистрации, страницы профиля.
Реализовать динамический маршрут для деталей ингредиента.

AJIEKCEY committed:
sprint-3/step-1
Добавить слайсы для авторизации, регистрации, страницы профиля.
Реализовать динамический маршрут для деталей ингредиента.

AJIEKCEY committed:
sprint-3/step-1
Добавить компонент для ошибок
Вынести дублирующий код форм в хук

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -0,0 +1,79 @@
+
+export const LoginPage = (): React.JSX.Element => {
+	const [email, setEmail] = useState('');
+	const [password, setPassword] = useState('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

AJIEKCEY replied:
Спасибо за идею. Добавил. Пока без валидации.

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -0,0 +1,79 @@
+
+	return (
+		<main className={styles.wrapper}>
+			<form className={styles.form} onSubmit={handleSubmit}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/pages/home/home.tsx:
```diff
@@ -0,0 +1,67 @@
+	// Загружаем ингредиенты при монтировании
+	useEffect(() => {
+		if (ingredients.length === 0) {
+			dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/ingredient/ingredient.tsx:
```diff
@@ -0,0 +1,59 @@
+	useEffect(() => {
+		// Если ингредиенты еще не загружены, загружаем их
+		if (ingredients.length === 0) {
+			dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

AJIEKCEY replied:
Не согласен. Так как есть страницы куда пользователь может попасть по прямой ссылке и на них не используется список ингредиентов то и нет необходимости их загружать в app. А чтобы не загружать повторно я добавил проверку на наличие в сторе.

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,69 +1,172 @@
 
+	// Проверяем авторизацию при загрузке приложения
 	useEffect(() => {
-		dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

AJIEKCEY replied:
Комментарий выше. 
Также добавил проверку на авторизацию перед отправкой заказа и защиту на переход на страницу reset-password.

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,45 @@
+	// Если маршрут защищенный (anonymous=false) и пользователь НЕ авторизован
+	// редирект на логин с сохранением текущего пути
+	if (!anonymous && !isAuthenticated) {
+		return <Navigate to='/login' state={{ from: location }} replace />;
```
Хорошо сделан `ProtectedRouteElement`

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,69 +1,172 @@
+	return (
+		<>
+			{/* Основные маршруты - если есть background, показываем фоновую страницу */}
+			<Routes location={background || location}>
```
Отлично сделана логика отображения попапа и страницы  под ним

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку.  
* Хорошо сделан `ProtectedRouteElement`
* Отлично сделана логика отображения попапа и страницы  под ним


но есть некоторые недочеты:



* Неавторизованный пользователь **не должен** отправлять заказ на сервер. При клике на кнопку его нужно переводить на логин и после удачного логина обратно к заказу.
* `dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.
* Пользователь не должен **напрямую** попасть на маршрут `/reset-password`. На эту страницу **никакой** пользователь не должен попадать напрямую. Только после нажатия на `восстановить пароль` нужно делать переход на эту страницу. Для этого нужно сделать проверку, был ли пользователь на странице `forgot-password` перед тем, как попасть на `reset-password`

AJIEKCEY committed:
sprint-3/step-1
Добавить проверку авторизации перед отправкой заказа.

AJIEKCEY committed:
sprint-3/step-1
Защитить /reset-password от перехода не со страницы /forgot-password
Вынести компонент appRoutes в отдельную директорию

AJIEKCEY committed:
sprint-3/step-1
Добавить хук use-form для управления полями форм.

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -0,0 +1,92 @@
+import styles from './login.module.css';
+
+export const LoginPage = (): React.JSX.Element => {
+	const { values, handleChange } = useForm({
+		email: '',
+		password: '',
+	});
+
```
Если будет интересно, посмотрите, как можно сделать свой **хук валидации**. 

Этот код помещают в отдельный файл `useFormAndValidation.js` в папке `hooks` и импортируют функцию туда, где нужно валидировать:

```js

import {useState, useCallback} from 'react';

export function useFormAndValidation() {
  const [ values, setValues ] = useState({});
  const [ errors, setErrors ] = useState({});
  const [ isValid, setIsValid ] = useState(true);

  const handleChange = (e) => {
    const {name, value} = e.target
    setValues({...values, [name]: value });
    setErrors({...errors, [name]: e.target.validationMessage});
    setIsValid(e.target.closest('form').checkValidity());
  };



  const resetForm = useCallback((newValues = {}, newErrors = {}, newIsValid = false) => {
    setValues(newValues);
    setErrors(newErrors);
    setIsValid(newIsValid);
  }, [setValues, setErrors, setIsValid]);

  return { values, handleChange, errors, isValid, resetForm, setValues, setIsValid };
}


```

одной строчкой запускается вся валидация:

```js

  const {values, handleChange, errors, isValid, setValues, resetForm} = useFormAndValidation()


```

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
